Allow installation on PHP 8.1

cedric-anne opened the pull request:
|    Q          |   A
|-------------- | ------
| Documentation | no
| Bugfix        | no
| BC Break      | no
| New Feature   | yes
| RFC           | no
| QA            | no

### Description

Allow installation on PHP 8.1.

As discussed on #61 , a temporary polyfill was used in dev context to be able to test against PHP 8.1.

cedric-anne committed:
Allow installation on PHP 8.1

Signed-off-by: Cédric Anne <cedric.anne@gmail.com>

froschdesign commented on polyfill/Cache/Storage/Adapter/Memory.php:
```diff
@@ -0,0 +1,768 @@
+ * @see       https://github.com/laminas/laminas-cache-storage-adapter-memory for the canonical source repository
+ * @copyright https://github.com/laminas/laminas-cache-storage-adapter-memory/blob/master/COPYRIGHT.md
+ * @license   https://github.com/laminas/laminas-cache-storage-adapter-memory/blob/master/LICENSE.md New BSD License
+ */
```
This kind of DocBlock header is no longer needed and can be removed completely.

cedric-anne replied:
OK. I removed it.

Ocramius commented:
@boesing could you check this one, when possible? Unsure about the polyfill implications :D

froschdesign commented:
@Ocramius 
The related fix is already present in laminas-cache-storage-adapter-memory but not yet released:

https://github.com/laminas/laminas-cache-storage-adapter-memory/blob/2.0.x/src/Memory.php#L117-L122

Some more work on laminas-cache itself and the adapters is needed to provide full support of PHP 8.1.
This means we can wait or use the same solution with polyfills as in laminas-view.

Another option is to remove the caching feature from the translator and use PSR 16 via a decorator – _which is already planned for the next major version_. But that means some more work and time.

cedric-anne commented:
Ping @boesing

boesing commented:
Even tho this could work, upstream projects which rely on `laminas-cache` might be stuck on PHP 8.0 due to the fact that PHP 8.1 support in `laminas-cache` v2 is something I've just considered after realizing that there are so many projects which rely on `laminas-cache` and its `StorageFactory` quirk.

Since v3 of `laminas-cache` does not have that static factory magic anymore, I definitely have to think about an MVC module which polyfills the `StorageFactory`. **Maybe** another v2.14.0 release with a PHP 8.1 backport (which will also include several minor releases in `laminas-cache-storage-adapter*` components) is something I have to consider as well as I don't think that the upgrade path to v3 is doable without having breaks in all other components which do work with `laminas-cache` StorageFactory static factory.

----

Regarding the polyfill, if it makes tests green and its only a dev autoloader quirk, I don't really care. But this won't change anything regarding the `laminas-cache` topic.

froschdesign commented:
> Regarding the polyfill, if it makes tests green and its only a dev autoloader quirk, I don't really care. But this won't change anything regarding the `laminas-cache` topic.

It is definitely only for the completion of the tests and we must inform that the cache support is not ready for 8.1.

cedric-anne commented:
`laminas-cache` is just a suggested dependency. As PHP 8.1 is not allowed by its `composer.json` file, people will be informed of this lack of support when they will try to install/require it with composer on a PHP 8.1 platform.
Also, I guess on PHP 8.1 only deprecation warnings are triggered. These warning may be a pain in CI as they may result in tests failing when strictiness is pushed to its maximum, but except maybe on some specific adapters, cache should work as expected on PHP 8.1.

froschdesign commented:
I think we can simplify the update and skip the unit tests with cache for PHP 8.1.

@boesing What do you think?

boesing commented:
@froschdesign would be fine with both 👌🏻

cedric-anne committed:
Fix deprecated usage of null params in IntlDateFormatter::__construct()

Signed-off-by: Cédric Anne <cedric.anne@gmail.com>

cedric-anne commented:
Hi,

I removed the laminas-cache memory adapter polyfill and disable related tests on `PHP >= 8.1.0-dev`.

Also, I added a commit to fix deprecated usage of null arguments in `IntlDateFormatter::__construct()`. Indeed, even if PHP Doc indicates these parameters could be set to `null`, [PHP code says something else](https://github.com/php/php-src/blob/master/ext/intl/dateformat/dateformat.stub.php#L11).

Is there something else I should/can do ?

Regards

froschdesign commented:
@cedric-anne 
I think that I will be able to create a new version at the weekend, which will contain some more fixes.

Thanks for your help! ❤️

cedric-anne commented:
> Also, I added a commit to fix deprecated usage of null arguments in `IntlDateFormatter::__construct()`. Indeed, even if PHP Doc indicates these parameters could be set to `null`, [PHP code says something else](https://github.com/php/php-src/blob/master/ext/intl/dateformat/dateformat.stub.php#L11).

For this one, I created [a bug on PHP bugtracker](https://bugs.php.net/bug.php?id=81606), and I proposed [a PR on PHP source](https://github.com/php/php-src/pull/7640).
I do not really know if it would be accepted, but in this case, my second commit will be useless.

pascalheidmann commented:
Any update on this PR? Would love to see it fixed so I can fix https://github.com/mezzio/mezzio-laminasrouter/pull/9 which builds on laminas-i18n 8.1 support ;)

froschdesign commented on test/Translator/TranslatorTest.php:
```diff
@@ -125,6 +128,9 @@ public function testTranslationFromDifferentSourceTypes()
         $this->assertEquals('Nachricht 9', $translator->translate('Message 9')); //translation-more-de_DE.php
     }
 
+    /**
+     * @requires PHP < 8.1.0-dev
+     */
```
Please remove these comments and use the related cache storage factory:

```json
"laminas/laminas-cache": "^3.1.2",
"laminas/laminas-cache-storage-deprecated-factory": "^1.0.0",
"laminas/laminas-cache-storage-adapter-memory": "^2.0.0",
```

froschdesign commented on composer.json:
```diff
@@ -25,7 +25,7 @@
     },
     "require": {
-        "php": "^7.3 || ~8.0.0",
+        "php": "^7.3 || ~8.0.0 || ~8.1.0",
```
Version 7.3 can be removed because the support ends in 11 days.

froschdesign commented on composer.json:
```diff
@@ -25,7 +25,7 @@
+        "php": "^7.3 || ~8.0.0 || ~8.1.0",
         "ext-intl": "*",
         "laminas/laminas-stdlib": "^2.7 || ^3.0",
         "laminas/laminas-zendframework-bridge": "^1.0"
```
The package can be removed.

froschdesign reviewed (changes requested):
The new package [laminas-cache-storage-deprecated-factory](https://github.com/laminas/laminas-cache-storage-deprecated-factory) can be used to fix the problem with the deprecated storage factory.

cedric-anne committed:
Remove laminas/laminas-zendframework-bridge dependency

Signed-off-by: Cédric Anne <cedric.anne@gmail.com>

Ocramius reviewed (changes requested):
```
 > - Downloading squizlabs/php_codesniffer (2.9.2)
```

This is *very* old, and likely the reason why you are getting failures on CS checks :thinking:

froschdesign commented:
@Ocramius 
> This is _very_ old, and likely the reason why you are getting failures on CS checks 🤔

The reason:

https://github.com/laminas/laminas-i18n/blob/fea1847eba42fabccedccfd3b12361196fe35bf5/composer.json#L35

cedric-anne commented:
> @Ocramius
> 
> > This is _very_ old, and likely the reason why you are getting failures on CS checks thinking
> 
> The reason:
> 
> https://github.com/laminas/laminas-i18n/blob/fea1847eba42fabccedccfd3b12361196fe35bf5/composer.json#L35

I'm working on it. Problem appears when dropping PHP 7.3 support. Maybe it should be done in another PR, as there are many CS fixes to do to be able to upgrade `laminas/laminas-coding-standard`.

froschdesign commented:
@cedric-anne 
> Maybe it should be done in another PR, as there are many CS fixes to do to be able to upgrade `laminas/laminas-coding-standard`.

If it helps here then no problem! 👍

cedric-anne commented:
It is OK now. I upgraded all dev dependencies to versions compatible with PHP `^7.3 || ~8.0.0` or `^7.3 || ~8.0.0 || ~8.1.0` and used the `--prefer-lowest` flag to ensure that all of them are retrieved in their PHP 7.3 compatible version.

PS: sorry for the force-push mess.

froschdesign commented:
@cedric-anne 
Thank you for your time and this contribution! 👍
